core/src/trezor/ui/text.py

```diff
@@ -15,109 +15,255 @@
+            length = len(string) - start
+
+        self.start = start
+        self.length = length
+        self.width = result_width
+        self.word_break = word_break
+        self.advance_whitespace = advance_whitespace
+        return start + length < len(string)
+
+
+_WORKING_SPAN = Span()
+
+
 def render_text(
-    words: List[TextContent],
+    items: List[TextContent],
     new_lines: bool,
     max_lines: int,
     font: int = ui.NORMAL,
     fg: int = ui.FG,
     bg: int = ui.BG,
     offset_x: int = TEXT_MARGIN_LEFT,
     offset_y: int = TEXT_HEADER_HEIGHT + TEXT_LINE_HEIGHT,
-    offset_x_max: int = ui.WIDTH,
+    line_width: int = ui.WIDTH - TEXT_MARGIN_LEFT,
+    item_offset: int = 0,
+    char_offset: int = 0,
+    break_words: bool = False,
+    render_page_overflow: bool = True,
 ) -> None:
+    """Render a sequence of items on screen.
+
+    The items can either be strings, or rendering instructions specified as ints.
+    They can change font, insert an explicit linebreak, or change color of the following
+    text.
+
+    If `new_lines` is true, a linebreak is rendered after every string. In effect, the
+    following calls are equivalent:
+
+    >>> render_text(["hello", "world"], new_lines=True)
+    >>> render_text(["hello\nworld"], new_lines=False)
+
+    TODO, we should get rid of all cases that use `new_lines=True`
+
+    If the rendered text ends up longer than `max_lines`, a trailing "..." is rendered
+    at end. This indicates to the user that the full contents have not been shown.
+    It is possible to override this behavior via `render_page_overflow` argument --
+    if false, the trailing "..." is not shown. This is useful when the rendered text is
+    in fact paginated.
+
+    `font` specifies the default font, but that can be overriden by font instructions
+    in `items`.
+    `fg` specifies default foreground color, which can also be overriden by instructions
+    in `items`.
+    `bg` specifies background color. This cannot be overriden.
+
+    `offset_x` and `offset_y` specify starting XY position of the text bounding box.
+    `line_width` specifies width of the bounding box. Height of the bounding box is
+    calculated as `max_lines * TEXT_LINE_HEIGHT`.
+
+    `item_offset` and `char_offset` must be specified together. Item offset specifies
+    the first element of `items` which should be considered, and char offset specifies
+    the first character of the indicated item which should be considered.
+    The purpose is to allow rendering different "pages" of text, using the same `items`
+    argument (slicing the list could be expensive in terms of memory).
+    The item selected by `item_offset` must be a string.
+
+    If `break_words` is false (default), linebreaks will only be rendered (a) at
+    whitespace, or (b) in case a word does not fit on a single line. If true, whitespace
+    is ignored and linebreaks are inserted after the last character that fits.
+    """
```

Comment:
Ran into this corner case where `\n` gets printed:
![emu00000000](https://user-images.githubusercontent.com/85857/102286324-5ca57d80-3f38-11eb-81ad-4a9f4394a2b7.png)
```
text = Text("title", new_lines=False)
text.normal("1234567890 1234567", "\nfoo")
return interact(ctx, Confirm(text), "test")
```
Was trying to write naive reimplementation of `format_parametrized` and this sometimes happened with `foo\n{}\nbar` format.

Reply:
nice catch! this is because of the `XXX hack` part which tries to render the string as a whole if it fits on a line by itself.

Wondering what to do about it. I want to get rid of the whole branch, but that introduces too many UI diffs at the moment that would need to be fixed manually.

OTOH if we wait with merging this until we have the redesign ready, we can do exactly that: remove the hack branch and make sure no layouts are relying on that behavior.

Is it a problem for you to patch out the hack branch and continue working on that version? Or do you also rely on the good UI results?
In any case this *seems* fixable by adding `and "\n" not in item` to the hack condition?

Reply:
Currently I'm aiming for as little UI changes as possible, however removing the branch after redesign is ready works for me, will work around it in the meantime.

Reply:
Can we just modify the branch to do a newline and then go on with rendering as usual? PTAL @ 3d375e86 Should produce no [UI diff](https://gitlab.com/satoshilabs/trezor/trezor-firmware/-/jobs/920002077).

Reply:
yes, this seems reasonable. I'll cherry-pick that commit

Reply:
OTOH it seems that in your case it will cause an additional newline, right? that's not good.

Reply:
Yes, good point, didn't realize this and no UI-recorded test triggers it. Would it be too much of a hack to detect the case and eat the `\n`?
```python
        item_width = ui.display.text_width(item, font)
        if (
            item_width <= line_width
            and item_width + offset_x - INITIAL_OFFSET_X > line_width
        ):
            offset_x = INITIAL_OFFSET_X
            offset_y += TEXT_LINE_HEIGHT
            if len(item) > char_offset and item[char_offset] == "\n":
                char_offset += 1

```

Reply:
I'd prefer if we avoided hitting the hack branch at all in new code. We don't want to create layouts that rely on its existence.
Adding `and "\n" not in item` to the hack condition will fully skip it for new-style strings, which I believe is acceptable.

Reply:
Works for me!

Reply:
resolved in 82ce9be47460baa8a3f4857f3f9ff9170a1ce95c
sorry for not cherry-picking, i had some git mess to clean here and the cherrypick commit disappeared in the process :(